finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+        let missing_authorities = handler.missing_nodes();
+        if missing_authorities.is_empty() {
+            return Vec::new();
+        }
+        let node_count = handler.node_count();
+        let authentication = handler.authentication();
+        if missing_authorities.len() * 3 > 2 * node_count.0 {
+            // We know of fewer than 1/3 authorities, broadcast our authentication and hope others
+            // respond in kind.
+            vec![authentication_broadcast(authentication)]
+        } else {
+            // Attempt learning about more authorities from the ones you already know.
+            let mut result = Vec::new();
+            let mut target = self.next_query;
+            while result.len() < 2 {
+                if let Some(peer_id) = handler.peer_id(&target) {
+                    result.push(request(
+                        missing_authorities.clone(),
+                        authentication.clone(),
+                        peer_id,
+                    ));
+                }
+                target = NodeIndex((target.0 + 1) % node_count.0);
+            }
+            self.next_query = target;
+            result
+        }
+    }
+
+    fn handle_authentication(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> Vec<Multiaddr> {
+        if !handler.handle_authentication(authentication.clone()) {
+            return Vec::new();
+        }
+        authentication.0.address()
+    }
+
+    fn handle_broadcast(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>) {
+        let addresses = self.handle_authentication(authentication.clone(), handler);
+        if addresses.is_empty() {
+            return (addresses, None);
+        }
+        let node_id = authentication.0.creator();
+        let rebroadcast = match self.last_broadcast.get(&node_id) {
+            Some(instant) => time::Instant::now() > *instant + self.cooldown,
+            None => true,
+        };
+        let message = match rebroadcast {
+            true => {
+                self.last_broadcast.insert(node_id, time::Instant::now());
+                Some(authentication_broadcast(authentication))
+            }
+            false => None,
+        };
+        (addresses, message)
+    }
+
+    fn create_response(
+        &mut self,
+        requester_id: NodeIndex,
+        node_ids: Vec<NodeIndex>,
+        handler: &mut SessionHandler,
+    ) -> Option<(DiscoveryMessage, DataCommand)> {
+        let requested_authentications = self
+            .requested_authentications
+            .entry(requester_id)
+            .or_default();
+        requested_authentications.extend(node_ids);
+        if let Some(instant) = self.last_response.get(&requester_id) {
+            if time::Instant::now() < *instant + self.cooldown {
```

Comment:
just thinking: is a common cooldown for both broadcast and direct messages appropriate? probably yes :thinking:

Reply:
Lets say it's not inappropriate. <_<"

They could be separate, but I couldn't think of any reason we would make them such.